### PR TITLE
KEP-1757 - client support for swarm busy errors

### PR DESCRIPTION
**Pull request by paularchard**
and other assorted non-database errors.

### CODE REVIEW

File: proto/bluzelle.proto [hunk 1]
```diff
@@ -1,5 +1,12 @@
 syntax = "proto3";
 
+message swarm_error
```
Comment (ebruck) [inline proto/bluzelle.proto:3]: Does this need to stay insync with swarmdb at all?

File: node/node.cpp [hunk 2]
```diff
@@ -18,12 +18,17 @@
 
 using namespace bzapi;
 
+namespace
+{
+    uint64_t MAX_BACKOFF_TIME{1024};
```
Comment (ebruck) [inline node/node.cpp]: should be const?

File: node/node.cpp [hunk 3]
```diff
@@ -154,27 +172,49 @@ node::queued_send(const std::string& msg, bzn::asio::write_handler callback)
 }
 
 void
-node::do_send()
+node::send_func()
```
Comment (ebruck) [inline node/node.cpp]: func is sort of redundant?

File: node/node.cpp [hunk 4]
```diff
@@ -154,27 +172,49 @@ node::queued_send(const std::string& msg, bzn::asio::write_handler callback)
...
+}
+
+void
+node::do_send()
+{
+    if (this->backoff_time)
+    {
+        this->backoff_timer->expires_from_now(std::chrono::milliseconds{this->backoff_time});
+        this->backoff_timer->async_wait(this->strand->wrap([weak_this = weak_from_this()](const auto& /*ec*/)
+        {
```
Comment (ebruck) [inline node/node.cpp]: start lambda on next line

File: node/node.cpp [hunk 5]
```diff
@@ -154,27 +172,49 @@ node::queued_send(const std::string& msg, bzn::asio::write_handler callback)
...
+            if (auto strong_this = weak_this.lock())
```
Comment (ebruck) [inline node/node.cpp]: do we need to check ec?

File: database/db_dispatch.cpp [hunk 6]
```diff
@@ -164,6 +165,66 @@ db_dispatch::handle_swarm_response(const bzn_envelope& response)
...
+        return true;
+    }
+
+    auto i = this->messages.find(nonce);
+    if (i == this->messages.end())
+    {
+        LOG(trace) << "Ignoring error response for unknown or already processed message: " << nonce;
+        return false;
+    }
+    auto& info = i->second;
```
Comment (ebruck) [inline database/db_dispatch.cpp]: do you need this way up here? It's not used until #221

**Review by ebruck — CHANGES_REQUESTED** [review]
just some minor points

File: proto/bluzelle.proto (hunk 1 above)
Comment (paularchard) [reply proto/bluzelle.proto:3]: Yes it should be kept the same in both projects. We already have that issue. Ideally we would have stuff like this in an external project included in both projects.

Comment (paularchard) [reply proto/bluzelle.proto:3]: Having said that, if the swarmdb we're talking to hasn't been updated to send these messages yet, it won't be a problem.

File: node/node.cpp (hunk 2 above)
Comment (paularchard) [reply node/node.cpp]: done

File: node/node.cpp (hunk 3 above)
Comment (paularchard) [reply node/node.cpp]: renamed

File: node/node.cpp (hunk 4 above)
Comment (paularchard) [reply node/node.cpp]: done

File: node/node.cpp (hunk 5 above)
Comment (paularchard) [reply node/node.cpp]: done

File: database/db_dispatch.cpp (hunk 6 above)
Comment (paularchard) [reply database/db_dispatch.cpp]: done

**Review by ebruck — APPROVED** [review]

**Pull request merged**